source/Calamari.Aws/Integration/CloudFormation/CloudFormationObjectExtensions.cs

```diff
@@ -0,0 +1,327 @@
+            } while (clientFactory.StackExists(stack, StackStatus.Completed) == StackStatus.InProgress);
+        }
+                
+        /// <summary>
+        /// Check the stack event status to detemrine whether it was successful.
+        /// http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/using-cfn-describing-stacks.html#w2ab2c15c15c17c11
+        /// </summary>
+        /// <param name="status">The status to check</param>
+        /// <returns>true if the status indcates a failed create or update, and false otherwise</returns>
+        public static Maybe<bool> MaybeIndicatesSuccess(this StackEvent status)
+        {
+            return status.ToMaybe().Select(x => !UnsuccessfulStackEvents.Contains(x.ResourceStatus.Value));
+        }
+
+        public static bool StackIsUnrecoverable(this StackEvent status)
+        {
+            Guard.NotNull(status, "Status should not be null");
+            return UnrecoverableStackStatuses.Contains(status.ResourceStatus.Value);
+        }
+
+        public static DeleteStackResponse DeleteStack(this Func<IAmazonCloudFormation> clientFactory, StackArn stack)
+        {
+            Guard.NotNull(clientFactory, "clientFactory should not be null");
+            Guard.NotNull(stack, "Stack should not be null");
+
+            try
+            {
+                return clientFactory().Map(client =>
+                    client.DeleteStack(new DeleteStackRequest {StackName = stack.Value}));
+            }
+            catch (AmazonCloudFormationException ex)
+            {
+                if (ex.ErrorCode == "AccessDenied")
+                {
+                    throw new PermissionException(
+                        "AWS-CLOUDFORMATION-ERROR-0009: The AWS account used to perform the operation does not have " +
+                        "the required permissions to delete the stack.\n" +
+                        ex.Message + "\n" +
+                        "For more information visit https://g.octopushq.com/AwsCloudFormationDeploy#aws-cloudformation-error-0009");
+                }
+
+                throw new UnknownException(
+                    "AWS-CLOUDFORMATION-ERROR-0010: An unrecognised exception was thrown while deleting a CloudFormation stack.\n" +
+                    "For more information visit https://g.octopushq.com/AwsCloudFormationDeploy#aws-cloudformation-error-0010",
+                    ex);
+            }
+        }
+
+        public static string CreateStack(this Func<IAmazonCloudFormation> clientFactory,
+            CreateStackRequest request)
+        {
+            try
+            {
+                return clientFactory()
+                    .Map(client => client.CreateStack(request))
+                    // Narrow the response to the stack ID
+                    .Map(response => response.StackId);
+            }
+            catch (AmazonCloudFormationException ex)
+            {
+                if (ex.ErrorCode == "AccessDenied")
```

Comment:
Does AWS have an enum for these kinds of error codes rather than string matching? Probably not I'm guessing :)

Reply:
Unfortunately not... The only way we can basically deal with this without taking any more dependencies on things / going completely crazy is by catching and rethrowing as an Exception which doesn't have things hidden in some obscure property as a string.